src/paho/mqtt/client.py

```diff
@@ -2237,6 +2237,11 @@ def _message_retry_check(self):
     def _messages_reconnect_reset_out(self):
         with self._out_message_mutex:
             self._inflight_messages = 0
+            
+            if self._clean_session:
+                self._out_messages = []
```

Comment:
I don't think the standard impose to discard *all* messages. We must only discard message for which the publish packet was sent.
_out_messages contains all messages that will be send or are currently being send (for QoS > 0).
Excepted if i'm missing a part of the standard, I think we should keep not yet send messages. Without this, code like this one will start to fail:
```
client = mqtt.Client()
msg = client.publish('topic', qos=1)  # call publish *before* connection
client.connect("localhost", 1883, 60)
client.loop_start()
msg.wait_for_publish()  # this will hang forever if we drop all _out_messages
```

Reply:
Okay....That will work as well!!
Is this the correct way to implement this approach? Is it possible that message state(m.state) is mqtt_ms_queued(because we have to keep those messages as well)? I can't think of any case where message state is mqtt_ms_queued.

     def _messages_reconnect_reset_out(self):
        with self._out_message_mutex:
            self._inflight_messages = 0
            
            if self._clean_session:
                for m in self._out_messages:
                    m.timestamp = 0
                    if m.state != mqtt_ms_publish:
                        self._out_messages.pop(self._out_messages.index(m))
                return
    
            for m in self._out_messages:
                m.timestamp = 0
                if self._max_inflight_messages == 0 or self._inflight_messages < self._max_inflight_messages:
                    if m.qos == 0:
                        m.state = mqtt_ms_publish
                    elif m.qos == 1:
                        # self._inflight_messages = self._inflight_messages + 1
                        if m.state == mqtt_ms_wait_for_puback:
                            m.dup = True
                        m.state = mqtt_ms_publish
                    elif m.qos == 2:
                        # self._inflight_messages = self._inflight_messages + 1                           
                        if m.state == mqtt_ms_wait_for_pubcomp:
                            m.state = mqtt_ms_resend_pubrel
                            m.dup = True
                        else:
                            if m.state == mqtt_ms_wait_for_pubrec:
                                m.dup = True
                            m.state = mqtt_ms_publish
                else:
                    m.state = mqtt_ms_queued

Reply:
Yes I think this is a good solution. Would should also keep mqtt_ms_queued state.
mqtt_ms_publish happen when trying to send Qos > 0 and connection is not established
mqtt_ms_queued happen when we reach max_inflight_message.

In both case since publish packet were not sent, we can keep the message for the next session.